gapis/trace/android/trace.go

```diff
@@ -0,0 +1,487 @@
+						[]string{},
+						fmt.Sprintf("%s/%s", pkg, activity),
+						pkg,
+						"",
+					}, nil
+				}
+			}
+			return nil, log.Errf(ctx, nil, "Could not find Intent %s, in package %s/%s", intent, pkg, activity)
+		}
+		r := &tracer.TraceTargetTreeNode{
+			activity,
+			packages.GetIcon(instPkg),
+			fmt.Sprintf("%s/%s", pkg, activity),
+			"",
+			[]string{},
+			pkg,
+			pkg,
+			"",
+		}
+		var defaultAction *pkginfo.Action
+		for _, i := range act.Actions {
+			r.Children = append(r.Children, fmt.Sprintf("%s:%s/%s", i.Name, pkg, activity))
+			if i.IsLaunch {
+				defaultAction = i
+			}
+		}
+		if len(act.Actions) == 1 {
+			defaultAction = act.Actions[0]
+		}
+		if defaultAction != nil {
+			r.TraceURI = fmt.Sprintf("%s:%s/%s", defaultAction.Name, pkg, activity)
+		}
+		return r, nil
+	}
+
+	defaultAction := ""
+	r := &tracer.TraceTargetTreeNode{
+		pkg,
+		packages.GetIcon(instPkg),
+		pkg,
+		"",
+		[]string{},
+		"",
+		pkg,
+		"",
+	}
+	for _, a := range instPkg.Activities {
+		if len(a.Actions) > 0 {
+			r.Children = append(r.Children, fmt.Sprintf("%s/%s", pkg, a.Name))
+			for _, act := range a.Actions {
+				if act.IsLaunch {
+					defaultAction = fmt.Sprintf("%s:%s/%s", act.Name, pkg, a.Name)
+				}
+			}
+		}
+	}
+
+	if len(instPkg.Activities) == 1 {
+		if len(instPkg.Activities[0].Actions) == 1 {
+			defaultAction = fmt.Sprintf("%s:%s/%s", instPkg.Activities[0].Actions[0].Name, pkg, instPkg.Activities[0].Name)
+		}
+	}
+	r.TraceURI = defaultAction
+	return r, nil
+}
+
+// InstallPackage installs the given package onto the android device.
+// If it is a zip file that contains an apk and an obb file
+// then we install them seperately.
+// Returns a function used to clean up the package and obb
+func (t *androidTracer) InstallPackage(ctx context.Context, o *tracer.TraceOptions) (*android.InstalledPackage, func(), error) {
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return nil, nil, err
+	}
+	defer os.RemoveAll(tempDir)
+	// Call it .apk now, because it may actually be our apk
+	zipName := filepath.Join(tempDir + "zip.apk")
+	apkName := zipName
+	obbName := ""
+
+	if err = ioutil.WriteFile(zipName, o.UploadApplication, os.FileMode(0600)); err != nil {
+		return nil, nil, err
+	}
+
+	r, err := zip.OpenReader(zipName)
+	defer r.Close()
+	if err != nil {
+		return nil, nil, err
+	}
+	hasObb := false
+	if len(r.File) == 2 {
+		if (strings.HasSuffix(r.File[0].Name, ".apk") &&
+			strings.HasSuffix(r.File[1].Name, ".obb")) ||
+			(strings.HasSuffix(r.File[1].Name, ".apk") &&
+				strings.HasSuffix(r.File[0].Name, ".obb")) {
+			hasObb = true
+		}
+	}
+
+	if hasObb {
+		// We should extract the .zip file into a .apk and a .obb file
+		apkName = filepath.Join(tempDir, "a.apk")
+		obbName = filepath.Join(tempDir, "a.obb")
+
+		apkFile, err := os.OpenFile(apkName, os.O_RDWR|os.O_CREATE, 0600)
+		if err != nil {
+			return nil, nil, err
+		}
+		obbFile, err := os.OpenFile(obbName, os.O_RDWR|os.O_CREATE, 0600)
```

Comment:
Suggest moving this `if hasObb` block into a separate function so this mega function is reduced and using `defer obbFile.Close()`